sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -417,6 +417,14 @@ object NullPropagation extends Rule[LogicalPlan] {
         case left :: Literal(null, _) :: Nil => Literal.create(null, e.dataType)
         case _ => e
       }
+
+      // If the value expression is NULL then transform the In expression to
+      // Literal(null)
+      case e @ In(v, list) => v match {
```

Comment:
We will group several rules into one batch, and execute one batch many times. So the order doesn't matter, it may not match at the beginning, but hopelly it can match in the next execution.